src/rpc/evo.cpp

```
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral type not supported: %s", ptx.collateralOutpoint.ToStringShort()));
             }
 
-            CKey key;
-            if (!spk_man->GetKey(ToKeyID(*pkhash), key)) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+            if (isPrepareRegister) {
+                // external signing with collateral key
+                ptx.vchSig.clear();
+                SetTxPayload(tx, ptx);
+
+                UniValue ret(UniValue::VOBJ);
+                ret.pushKV("tx", EncodeHexTx(CTransaction(tx)));
+                ret.pushKV("collateralAddress", EncodeDestination(txDest));
+                ret.pushKV("signMessage", ptx.MakeSignString());
+                return ret;
+            } else {
+                // lets prove we own the collateral
+                LegacyScriptPubKeyMan* spk_man = wallet->GetLegacyScriptPubKeyMan();
+                if (!spk_man) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "This type of wallet does not support this command");
+                }
+
+                CKey key;
+                if (!spk_man->GetKey(ToKeyID(*pkhash), key)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+                }
+                SignSpecialTxPayloadByString(tx, ptx, key);
+                SetTxPayload(tx, ptx);
+                return SignAndSendSpecialTx(request, chainman, tx, fSubmit);
+            }
+        } catch (...) {
+            if (unlockOnError) {
+                LOCK(wallet->cs_wallet);
+                wallet->UnlockCoin(ptx.collateralOutpoint);
             }
-            SignSpecialTxPayloadByString(tx, ptx, key);
-            SetTxPayload(tx, ptx);
-            return SignAndSendSpecialTx(request, chainman, tx, fSubmit);
+            throw;
```

why is there a naked throw?

---

bcs need to re-throw the exception which has been caught by `except` without changing logic of code. If throw is naked - it throws the last exception again.

But I'd suggest to write here a wrapper around Lock/Unlock same as 'lock_guard + mutex works'.